test/utils.js

```diff
@@ -1,83 +1,118 @@
+      onDisconnect: _onDisconnect
     }
-  })
+
+    return multicodecs[0]
+  },
+  unregister: () => {}
+})
+
+// as created by libp2p
+exports.createPubsubNode = async (registrarRecord) => {
+  const peerInfo = await createPeerInfo()
+  peerInfo.protocols.add(multicodec)
+  const pubsub = new Pubsub(peerInfo, createMockRegistrar(registrarRecord))
+
+  await pubsub.start()
+
+  return {
+    peerInfo: pubsub.peerInfo,
+    subscribe: (topic, handler) => {
+      pubsub.subscribe(topic)
+
+      pubsub.on(topic, handler)
+    },
+    unsubscribe: (topic, handler) => {
+      if (!handler) {
+        pubsub.removeAllListeners(topic)
+      } else {
+        pubsub.removeListener(topic, handler)
+      }
+
+      pubsub.unsubscribe(topic)
+    },
+    publish: (topic, data) => pubsub.publish(topic, data),
+    getTopics: () => pubsub.getTopics(),
+    getSubscribers: (topic) => pubsub.getSubscribers(topic),
+    stop: () => pubsub.stop()
+  }
 }
 
-// stop a daemon
-const stopDaemon = async (daemon) => {
-  await daemon.stop()
-  await new Promise((resolve) => setTimeout(() => resolve(), 200))
-  return daemon.cleanup()
+const ConnectionPair = () => {
+  const [d0, d1] = DuplexPair()
+
+  return [
+    {
+      stream: d0,
+      newStream: () => Promise.resolve({ stream: d0 })
+    },
+    {
+      stream: d1,
+      newStream: () => Promise.resolve({ stream: d1 })
+    }
+  ]
 }
 
-// connect two peers
-const connect = (dA, dAId, dB, dBId) => {
-  const dALocalAddress = dAId.addresses.find(a => a.includes('127.0.0.1'))
-  const dBLocalAddress = dBId.addresses.find(a => a.includes('127.0.0.1'))
+exports.connectPubsubNodes = async (pubsubA, pubsubB) => {
+  const onConnectA = pubsubA.registrar[multicodec].onConnect
+  const onConnectB = pubsubB.registrar[multicodec].onConnect
 
-  return Promise.all([
-    dA.api.swarm.connect(dBLocalAddress),
-    dB.api.swarm.connect(dALocalAddress)
-  ])
+  // Notice peers of connection
+  const [c0, c1] = ConnectionPair()
+  await onConnectA(pubsubB.router.peerInfo, c0)
+  await onConnectB(pubsubA.router.peerInfo, c1)
 }
 
 // Wait for a condition to become true.  When its true, callback is called.
-const waitFor = async (predicate) => {
-  for (let i = 0; i < 10; i++) {
+exports.waitFor = (predicate) => {
+  return pWaitFor(async () => {
     if (await predicate()) {
-      return
+      return true
     }
 
-    await delay(1000)
-  }
-
-  throw new Error('waitFor time expired')
+    return false
+  }, {
+    interval: 1000,
+    timeout: 10000
+  })
 }
 
 // Wait until a peer subscribes a topic
-const waitForPeerToSubscribe = async (topic, peer, daemon) => {
-  for (let i = 0; i < 5; i++) {
-    const peers = await daemon.api.pubsub.peers(topic)
+exports.waitForPeerToSubscribe = (topic, peer, node) => {
+  return pWaitFor(async () => {
+    const peers = await node.getSubscribers(topic)
 
-    if (peers.includes(peer.id)) {
-      return
+    if (peers.includes(peer._idB58String)) {
```

Comment:
This will disappear soon, it'll be difficult to know if the `toB58String` function is being used or not if we use the private var.

```suggestion
    if (peers.includes(peer.id.toB58String())) {
```